mass-deployer/pkg/mass-deployer/deployer.go

```diff
@@ -0,0 +1,283 @@
+	}
+	return deploymentsInfo
+}
+
+func massDeploy(tfPluginClient deployer.TFPluginClient, ctx context.Context, deployemnts groupDeploymentsInfo) ([]string, error) {
+	err := tfPluginClient.NetworkDeployer.BatchDeploy(ctx, deployemnts.networkDeployments)
+	if err != nil {
+		return []string{}, err
+	}
+
+	err = tfPluginClient.DeploymentDeployer.BatchDeploy(ctx, deployemnts.vmDeployments)
+	if err != nil {
+		return []string{}, err
+	}
+	vmsInfo := loadDeploymentsInfo(tfPluginClient, deployemnts.deploymentsInfo)
+
+	return vmsInfo, nil
+}
+
+func buildDeployments(tfPluginClient deployer.TFPluginClient, vms []Vms, nodesIDs []int, sshKeys map[string]string) groupDeploymentsInfo {
+	var vmDeployments []*workloads.Deployment
+	var networkDeployments []*workloads.ZNet
+	var deployemntsInfo []vmDeploymentInfo
+	nodesIDsIdx := 0
+
+	// here we loop over all groups of vms within the same node group, and for every group
+	// we loop over all it's vms and create network and vm deployment for it
+	// the nodesIDsIdx is a counter used to get nodeID to be able to distribute load over all nodes
+	for _, vmGroup := range vms {
+		for i := 0; i < int(vmGroup.Count); i++ {
+			nodeID := uint32(nodesIDs[nodesIDsIdx%len(nodesIDs)])
+			nodesIDsIdx++
+
+			disks, mounts := parseDisks(vmGroup.Name, vmGroup.SSDDisks)
+
+			network := workloads.ZNet{
+				Name:        fmt.Sprintf("%s%dnetwork", vmGroup.Name, i),
+				Description: "network for mass deployment",
+				Nodes:       []uint32{nodeID},
+				IPRange: gridtypes.NewIPNet(net.IPNet{
+					IP:   net.IPv4(10, 20, 0, 0),
+					Mask: net.CIDRMask(16, 32),
+				}),
+				AddWGAccess: false,
+			}
+			w := workloads.VM{
+				Name:        fmt.Sprintf("%s%d", vmGroup.Name, i),
+				NetworkName: network.Name,
+				Flist:       vmGroup.Flist,
+				CPU:         int(vmGroup.FreeCPU),
+				Memory:      int(vmGroup.FreeMRU),
+				PublicIP:    vmGroup.Pubip4,
+				PublicIP6:   vmGroup.Pubip6,
+				Planetary:   vmGroup.Planetary,
+				RootfsSize:  int(convertGBToBytes(vmGroup.Rootsize)),
+				Entrypoint:  vmGroup.Entrypoint,
+				EnvVars:     map[string]string{"SSH_KEY": sshKeys[vmGroup.SSHKey]},
+				Mounts:      mounts,
+			}
+			deployment := workloads.NewDeployment(w.Name, nodeID, "", nil, network.Name, disks, nil, []workloads.VM{w}, nil)
+
+			vmDeployments = append(vmDeployments, &deployment)
+			networkDeployments = append(networkDeployments, &network)
+			deployemntsInfo = append(deployemntsInfo, vmDeploymentInfo{nodeID: nodeID, deploymentName: deployment.Name, vmName: w.Name})
+		}
+	}
+	return groupDeploymentsInfo{vmDeployments: vmDeployments, networkDeployments: networkDeployments, deploymentsInfo: deployemntsInfo}
+}
+
+func loadDeploymentsInfo(tfPluginClient deployer.TFPluginClient, deployemnts []vmDeploymentInfo) []string {
+	vmsInfo := []string{}
+	for _, info := range deployemnts {
+		vm, err := tfPluginClient.State.LoadVMFromGrid(info.nodeID, info.vmName, info.deploymentName)
+		if err != nil {
```

Comment:
can this be done concurrently? we don't need to wait on each vm. and as far as I know the node client supports concurrent requests.